### PR TITLE
Server certificate support for API 1200,1000,800

### PR DESCRIPTION
**Pull request by nabhajit-ray**
Server certificate support for API 1200,1000,800

### CODE REVIEW

File: oneview/resource_server_certificate_test.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// (C) Copyright 2016 Hewlett Packard Enterprise Development LP
```
Comment (jyothis-gireesan) [inline oneview/resource_server_certificate_test.go:1]: Copyright year

File: oneview/resource_server_certificate.go [hunk 2]
```diff
@@ -0,0 +1,533 @@
+// (C) Copyright 2016 Hewlett Packard Enterprise Development LP
```
Comment (jyothis-gireesan) [inline oneview/resource_server_certificate.go:1]: Copyright

File: docs/d/hypervisor_manager.html.markdown [hunk 3]
```diff
@@ -0,0 +1,64 @@
+---
+layout: "oneview"
+page_title: "Oneview: hypervisor_manager"
+sidebar_current: "docs-hypervisor_manager"
+description: |-
+ Gets information about an existing hypervisor_manager.
+---
```
Comment (AnushaY1916) [inline docs/d/hypervisor_manager.html.markdown:7]: This file belongs to Hypervisor, please remove this.

File: docs/d/server_certificate.html.markdown [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+output "oneview_server_certificate_value" {
+ value = "${data.oneview_server_certificate.test.uri}"
+}
+```
+
+## Argument Reference
+
+* `alias_name` - (Required) The alias name of an existing server certificate.
```
Comment (AnushaY1916) [inline docs/d/server_certificate.html.markdown:27]: Making both attributes required is not correct. Describe correctly that any one of these attributes can be used to get the Server certificate.

File: docs/r/hypervisor_manager.html.markdown [hunk 5]
```diff
@@ -0,0 +1,73 @@
+---
+layout: "oneview"
+page_title: "Oneview: hypervisor_manager"
+sidebar_current: "docs-oneview-hypervisor-manager"
+description: |-
+  Creates a hypervisor manager.
```
Comment (AnushaY1916) [inline docs/r/hypervisor_manager.html.markdown:6]: Remove this file too.

File: docs/r/server_certificate.html.markdown [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+* `type` - (Required) Indicates the type of the Server certificate.
+* `base64_data` - (Required) Encrypted content of the SSL certificate.
+
+## Attributes Reference
+
+In addition to the arguments listed above, the following computed attributes are exported:
+
+* `alternative_name` -  An optional entry that contains additional names that apply to the owner of the certificate, possibly including additional email addresses, DNS names, IP addresses, or other identifiers.
+
+* `base64_data` -  Encrypted content of the SSL certificate.
```
Comment (AnushaY1916) [inline docs/r/server_certificate.html.markdown:35]: This attribute is already mentioned above. Remove this duplicate.

File: docs/r/server_certificate.html.markdown [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+# oneview\_server\_certificate
+
+Creates an server certificate.
+
+## Example Usage
+
+```js
+resource "oneview_server_certificate" "default" {
+  name = "test-server-certificate"
```
Comment (AnushaY1916) [inline docs/r/server_certificate.html.markdown:17]: This request is not complete. Please mention the request with the required attributes.

File: examples/hypervisor_manager.tf [hunk 8]
```diff
@@ -0,0 +1,48 @@
+provider "oneview" {
+	ov_username = "Administrator"
```
Comment (AnushaY1916) [inline examples/hypervisor_manager.tf:2]: Remove this file.

File: examples/server_certificate.tf [hunk 9]
```diff
@@ -0,0 +1,60 @@
+provider "oneview" {
+	ov_username = "Administrator"
+	ov_password = "admin123"
```
Comment (AnushaY1916) [inline examples/server_certificate.tf:3]: Replace actual values with the placeholders.

File: examples/server_certificate.tf [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	ov_apiversion = 1200
+	ov_ifmatch = "*"
+}
+
+data "oneview_scope" "scope_obj" {
+        name = "test_scope"
+}
+
+resource "oneview_server_certificate" "ServerCertificate" {
+
```
Comment (AnushaY1916) [inline examples/server_certificate.tf:15]: File formatting is not proper. There are many empty lines and many tab spaces.

File: examples/server_certificate.tf [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+data "oneview_scope" "scope_obj" {
+        name = "test_scope"
+}
+
+resource "oneview_server_certificate" "ServerCertificate" {
+                                  
+
+
+                                   certificate_details = [{
```
Comment (AnushaY1916) [inline examples/server_certificate.tf:18]: In the documentation, you have mentioned that name is mandatory but in examples, the attribute name is not given. Please check the REST API documentation and make it proper.

File: examples/server_certificate.tf [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+]
+
+read_type="name"
+                             
+}
+*/
+/* Testing data source
+
+data "oneview_server_certificate" "sc" {
```
Comment (AnushaY1916) [inline examples/server_certificate.tf:50]: Comment saying that any one of the attributes can be used for the data source.

File: oneview/data_source_hypervisor_manager.go [hunk 13]
```diff
@@ -0,0 +1,217 @@
+// (C) Copyright 2020 Hewlett Packard Enterprise Development LP
+//
```
Comment (AnushaY1916) [inline oneview/data_source_hypervisor_manager.go:2]: Remove this file

File: oneview/data_source_hypervisor_manager_test.go [hunk 14]
```diff
@@ -0,0 +1,40 @@
+// (C) Copyright 2019 Hewlett Packard Enterprise Development LP
```
Comment (AnushaY1916) [inline oneview/data_source_hypervisor_manager_test.go:1]: Remove this file.

File: oneview/data_source_server_certificate.go [hunk 15]
```diff
@@ -0,0 +1,368 @@
+// (C) Copyright 2016 Hewlett Packard Enterprise Development LP
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate.go:1]: Change the year

File: oneview/data_source_server_certificate.go [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"category": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"certificate_details": {
+				Type:     schema.TypeSet,
+				Optional: true,
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate.go:34]: Make this Computed

File: oneview/data_source_server_certificate.go [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"alias_name": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate.go:39]: Make this Computed

File: oneview/data_source_server_certificate.go [hunk 18]
```diff
@@ -0,0 +1,368 @@
...
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"surname": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate.go:187]: Make this Computed

File: oneview/data_source_server_certificate.go [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+			},
+
+			"certificate_status": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"chain_status": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate.go:216]: Make this Computed

File: oneview/data_source_server_certificate.go [hunk 20]
```diff
@@ -0,0 +1,368 @@
...
+			},
+			"etag": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"modified": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate.go:245]: Make this Computed. Attributes  alias_name and remote_ip are used to make read call here, so except these values make rest of the values as Computed.

File: oneview/data_source_server_certificate_test.go [hunk 21]
```diff
@@ -0,0 +1,40 @@
+// (C) Copyright 2019 Hewlett Packard Enterprise Development LP
```
Comment (AnushaY1916) [inline oneview/data_source_server_certificate_test.go:1]: Change the year.

File: oneview/provider.go [hunk 22]
```diff
@@ -72,12 +72,14 @@ func Provider() terraform.ResourceProvider {
 			"oneview_fcoe_network":               dataSourceFCoENetwork(),
 			"oneview_enclosure":                  dataSourceEnclosure(),
 			"oneview_ethernet_network":           dataSourceEthernetNetwork(),
+			"oneview_hypervisor_manager":         dataSourceHypervisorManager(),
```
Comment (AnushaY1916) [inline oneview/provider.go:75]: Remove this entry

File: oneview/provider.go [hunk 23]
```diff
@@ -96,12 +98,14 @@ func Provider() terraform.ResourceProvider {
 			"oneview_server_profile":             resourceServerProfile(),
 			"oneview_enclosure":                  resourceEnclosure(),
 			"oneview_enclosure_group":            resourceEnclosureGroup(),
+			"oneview_hypervisor_manager":         resourceHypervisorManager(),
```
Comment (AnushaY1916) [inline oneview/provider.go:101]: Remove this entry

File: oneview/resource_hypervisor_manager.go [hunk 24]
```diff
@@ -0,0 +1,364 @@
+// (C) Copyright 2020 Hewlett Packard Enterprise Development LP
```
Comment (AnushaY1916) [inline oneview/resource_hypervisor_manager.go:1]: Remove all the hypervisor related files.

File: oneview/resource_server_certificate.go [hunk 25]
```diff
@@ -0,0 +1,533 @@
...
+		return servCError
+	}
+	return resourceServerCertificateRead(d, meta)
+}
+
+func resourceServerCertificateRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	var servC ov.ServerCertificate
+	var err error
+	if val, ok := d.GetOk("read_type"); ok {
```
Comment (AnushaY1916) [inline oneview/resource_server_certificate.go:363]: Read type is always done by name and never by remote IP. There is no need for this read_type attribute. Read function in the resource file is called by either Create or Update and I see that in both the calls you are setting the ID with the alias_name. Read by remote IP is never used. So remove the read_type attribute, this if-else check, and call by remote IP, keep read by Name only.

File: oneview/resource_server_certificate.go [hunk 26]
```diff
@@ -0,0 +1,533 @@
...
+		servcCertificateStat := ov.CertificateStat{
+			ChainStatus: certificatestat["chain_status"].(string),
+			SelfSigned:  certificatestat["self_signed"].(bool),
+			Trusted:     certificatestat["trusted"].(bool),
+		}
+		servC.CertificateStatus = &servcCertificateStat
+	}
+
+	servCError := config.ovClient.UpdateServerCertificate(servC)
+	d.SetId(d.Get("name").(string))
```
Comment (AnushaY1916) [inline oneview/resource_server_certificate.go:518]: In create, Id is set with alias_name and here it's done with name. As per REST API documentation, GET calls are based on alias_name or IP address but not with a name. Please replace this name with alias_name and test this resource again.

**Pull request closed without merging**